shared/subspace-object-fetching/src/object_fetcher.rs

```diff
@@ -0,0 +1,402 @@
+// Copyright (C) 2024 Subspace Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Fetching objects stored in the archived state of Subspace Network.
+
+use async_trait::async_trait;
+use parity_scale_codec::{Compact, CompactLen, Decode, Encode};
+use std::fmt;
+use std::sync::Arc;
+use subspace_archiving::archiver::{NewArchivedSegment, Segment, SegmentItem};
+use subspace_core_primitives::objects::GlobalObject;
+use subspace_core_primitives::{
+    ArchivedHistorySegment, Piece, PieceIndex, RawRecord, RecordedHistorySegment, SegmentIndex,
+};
+use subspace_networking::utils::piece_provider::{PieceProvider, PieceValidator};
+use tracing::error;
+
+/// Maximum expected size of one object in bytes
+const MAX_OBJECT_SIZE: usize = 5 * 1024 * 1024;
+
+/// The maximum number of random walking rounds to fetch a piece from archival storage
+const MAX_RANDOM_WALK_ROUNDS: usize = 15;
+
+/// Top-level error type for this module.
+// TODO: More specific errors instead of `StringError`
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Errors that can be formatted as a String
+    #[error("{0}")]
+    StringError(String),
+}
+
+/// Something that can be used to get decoded pieces by index
+#[async_trait]
+pub trait ObjectPieceGetter: fmt::Debug {
+    /// Get piece by index
+    async fn get_piece(&self, piece_index: PieceIndex) -> Option<Piece>;
+}
+
+#[async_trait]
+impl<PG> ObjectPieceGetter for Vec<PG>
```

Comment:
What is the use-case for this implementation?

Reply:
You have a list of alternative ObjectPieceGetter implementations, for example, from memory (NewArchivedSegment), then from the local cache, then from the network.